param/parameters.py

```diff
@@ -2598,6 +2600,81 @@ def _validate_value(self, val, allow_None):
 #   - use resolve_path(path_to_file=False) for paths to existing folders to be read,
 #     and normalize_path() for paths to new files to be written.
 
+def _get_default_search_paths()->typing.Tuple[str]:
+    # Its 10x faster to pickle a string than a Path.
+    # Thus we use os.getcwd instead of pathlib.Path.cwd
+    return (os.getcwd(), )
+
+@lru_cache
+def _resolve_path_fast(
+    path: str,
+    search_paths: typing.Tuple[str|pathlib.Path],
+    path_to_file: bool|None=None
+)->str:
+
+    _path = pathlib.Path(os.path.normpath(path))
+    ftype = "File" if path_to_file is True \
+        else "Folder" if path_to_file is False else "Path"
+
+    if _path.is_absolute():
+        if ((path_to_file is None  and _path.exists()) or
+            (path_to_file is True  and _path.is_file()) or
+            (path_to_file is False and _path.is_dir())):
+            return str(_path)
+        raise OSError(f"{ftype} '{path}' was not found.")
+
+    paths_tried = []
+    for prefix in search_paths:
+        try_path = (pathlib.Path(prefix) / _path).resolve()
+
+        if ((path_to_file is None  and try_path.exists()) or
+            (path_to_file is True  and try_path.is_file()) or
+            (path_to_file is False and try_path.is_dir())):
+            return str(try_path.absolute())
+
+        paths_tried.append(str(try_path))
+
+    raise OSError(
+        f"{ftype} {_path.name} was not found in the following place(s): {str(paths_tried)}."
+    )
+
+def _resolve_path(
+    path: str,
+    search_paths: typing.List[str|pathlib.Path]|None=None,
+    path_to_file: bool|None=None
+)->str:
+    """
+    Find the path to an existing file, searching the paths specified
+    in the search_paths parameter if the filename is not absolute, and
+    converting a UNIX-style path to the current OS's format if
+    necessary.
+
+    To turn a supplied relative path into an absolute one, the path is
+    appended to paths in the search_paths parameter, in order, until
+    the file is found.
+
+    An IOError is raised if the file is not found.
+
+    Similar to Python's os.path.abspath(), except more search paths
+    than just os.getcwd() can be used, and the file must exist.
+
+    Parameters
+    ----------
+    search_paths : tuple, default=(os.getcwd(),)
+        Tuple of paths to search the path from
+    check_exists: boolean, default=True
+        If True (default) the path must exist on instantiation and set,
+        otherwise the path can optionally exist.
+    """
+    if not search_paths:
+        _search_paths = _get_default_search_paths()
+    else:
+        _search_paths = tuple(search_paths)
+    return _resolve_path_fast(path, _search_paths, path_to_file)
+
+def _get_default_search_paths_as_list()->typing.List[str]:
+    return [str(path) for path in _get_default_search_paths()]
+
 class resolve_path(ParameterizedFunction):
```

Comment:
Somewhere we should test users that there is a overhead cost of using this class instead of the `_resolve_path` function. I don't know if we should even deprecate it?